### PR TITLE
Add typecheck to split axis

### PR DESCRIPTION
**Pull request by delta2323**
Related #123

### CODE REVIEW

File: chainer/functions/split_axis.py [hunk 1]
```diff
@@ -26,6 +29,50 @@ def __init__(self, indices_or_sections, axis):
         self.indices_or_sections = indices_or_sections
         self.axis = axis
 
+    def check_type_forward(self, in_types):
+        type_check.expect(
+            in_types.size() == 1,
+            in_types[0].ndim >= self.axis
```
Comment (unnonouno) [inline chainer/functions/split_axis.py]: Don't use `in_types[0]` before you check `in_types.size() == 1`.

Comment (unnonouno) [inline chainer/functions/split_axis.py]: Why not use `Variable` for `self.axis`?

File: chainer/functions/split_axis.py [hunk 2]
```diff
@@ -26,6 +29,50 @@ def __init__(self, indices_or_sections, axis):
...
+        )
+
+        if isinstance(self.indices_or_sections, collections.Iterable):
+            max_index = type_check.Variable(
+                self.indices_or_sections[-1], 'max_index')
+            type_check.expect(in_types[0].shape[self.axis] >= max_index)
```
Comment (beam2d) [inline chainer/functions/split_axis.py]: Is this allowed to have the dimension equal to the `max_index`?

File: chainer/functions/split_axis.py [hunk 3]
```diff
@@ -26,6 +29,50 @@ def __init__(self, indices_or_sections, axis):
...
+    def check_type_backward(self, in_types, out_types):
+        if isinstance(self.indices_or_sections, collections.Iterable):
+            num_indices = type_check.Variable(
+                len(self.indices_or_sections), 'len(indices)')
+            output_size = num_indices + 1
+        else:
+            output_size = type_check.Variable(
+                self.indices_or_sections, 'sections')
+
+        type_check.expect(out_types.size() == output_size.eval())
```
Comment (unnonouno) [inline chainer/functions/split_axis.py]: You don't need to call `eval()`.

File: chainer/functions/split_axis.py [hunk 4]
```diff
@@ -26,6 +29,50 @@ def __init__(self, indices_or_sections, axis):
...
+
+        for out_type in out_types:
+            type_check.expect(
+                out_type.ndim == in_types[0].ndim,
+                out_type.dtype == out_types[0].dtype
```
Comment (unnonouno) [inline chainer/functions/split_axis.py]: `in_types[0].dtype`?

File: chainer/functions/split_axis.py [hunk 5]
```diff
@@ -26,6 +29,50 @@ def __init__(self, indices_or_sections, axis):
...
+        for out_type in out_types:
+            for i in six.moves.range(in_types[0].ndim.eval()):
+                if i == self.axis:
+                    continue
+                type_check.expect(out_type.shape[i] == in_types[0].shape[i])
+
+        total_output_dim = sum(
+            out_type.shape[self.axis] for out_type in out_types)
+        type_check.expect(
+            total_output_dim == in_types[0].shape[self.axis]
```
Comment (unnonouno) [inline chainer/functions/split_axis.py]: Do we need to check each value of `out_type.shape[self.axis]`?

File: chainer/functions/split_axis.py (hunk 2 above)
Comment (delta2323) [inline chainer/functions/split_axis.py]: I made a mistake. If the equality holds, last tensor of output is empty, which is forbidden.

File: chainer/functions/split_axis.py (hunk 3 above)
Comment (delta2323) [inline chainer/functions/split_axis.py]: As discussed in #209, `check_type_backward` is no longer necessary. So I remove this function.

File: chainer/functions/split_axis.py (hunk 4 above)
Comment (delta2323) [inline chainer/functions/split_axis.py]: I just intended to check that all out_types have same dtypes.

Comment (delta2323) [inline chainer/functions/split_axis.py]: As discussed in #209, `check_type_backward` is no longer necessary. So I remove this function.

File: chainer/functions/split_axis.py (hunk 5 above)
Comment (delta2323) [inline chainer/functions/split_axis.py]: I understand that the check is somewhat loose.

Comment (delta2323) [inline chainer/functions/split_axis.py]: As discussed in #209, `check_type_backward` is no longer necessary. So I remove this function.

**Comment by unnonouno** [comment]
LGTM

**Pull request merged**